### ISSUE TITLE
Panels: Disconnect Value Threshold does not interact well with automatic data point interval calculation

### ISSUE THREAD

**Issue by nagisa** [grafana/grafana#83031]
### What happened?

I have written a query that uses `$__rateInterval` in prometheus or `v.windowPeriod` in FluxV2. Graphed this in a Time Series graph. Knowing that my data is every 1 minute, I set the "Disconnect values" option to 5 minutes.

Now when I select a time range of say "Past 6 months", grafana picks an interval of 2 hours, which is more than the "disconnect values" setting and therefore every single point in the graph is now disconnected.

### What did you expect to happen?

I would expect there to be a way to specify this value in relation to the automatically computed interval. Of particular interest would be functions like:

* `max(5m, $__rateInterval)`
* `max(5m, $__rateInterval * 2)`

and similar.

### Did this work before?

Not that I know of.

### How do we reproduce it?

1. 6 months of data;
2. Downscale data according to the automatically computed `$__rateInterval` or `v.windowPeriod`;
3. Graph this as a timeseries
4. Set Disconnect Values > Threshold > 5m
5. Select a time range that computes interval greater than 5 minutes.

### Is the bug inside a dashboard panel?

Let me know if the instructions aren't clear, happy to share a panel then.

### Environment (with versions)?

Grafana: v10.3.1
OS: N/A
Browser: Firefox

### Grafana platform?

A package manager (APT, YUM, BREW, etc.)

### Datasource(s)?

InfluxDB, Prometheus

**Comment by leeoniya** [comment]
im not sure about allowing expressions here (it's complex), but i agree that internally our logic should probably ignore the disconnect setting when it's less than the `interval` (step) that comes back from the query. so if you set disconnect to 5m and step is computed to 2h, it will not disconnect every value.

so yes, `disconnect_threshold = max(disconnect_threshold, field.config.interval)` should be how it behaves by default. im not sure the influx datasource currently sets the Time field's `field.config.interval` in its query responses to `windowPeriod`. might be something we need to add for this to work seamlessly in influx. maybe influx returns `null` values in the periods without data, unlike prometheus which omits them.

**Comment by trueb2** [comment]
Adding expressions would solve the problem of requiring each data source implementation to support calculation of redundant null values. It would get rid of the need for queries like time_bucket_gapfill or similar that have additional performance or query constraints across the various data source backends.

Without expressions, the Disconnect Values by threshold concept doesn't make sense for time series data visualizations, where a dynamic grouping clashes with a static grouping threshold.

**Comment by Brascal** [comment]
I agree, this is a problem for graphs where A) it's important to show gaps in data collection, and B), it's a realistic use-case to want to query for both very long periods and very short periods. There is no disconnect threshold value that will work for all time ranges the user may select--without resulting in the graph disappearing! 

The only options I see at this point are to either forget about showing data collection gaps altogether, or creating a separate set of graphs dedicated to "long term" visualization, where the disconnect threshold is set to perhaps days or just eliminated altogether.

**Comment by leeoniya** [comment]
https://github.com/grafana/grafana/pull/88574 will try to improve this for influx by doing the same thing we do for Prometheus (set `field.config.interval`).

we have to be careful here, though, since Prometheus queries are typically aggregates and have "regular" time spacing, whereas i suspect influx queries are often irregular (arbitrary/raw IoT events) unless you request aggregation with a `windowPeriod` explicitly.

related: https://github.com/influxdata/flux/issues/2428

> There is no disconnect threshold value that will work for all time ranges the user may select--without resulting in the graph disappearing!

right, there isn't some magic way where we can tell for any datasource + zoom level + query + dataset whether we can assume regularly-spaced data to correctly auto-disconnect the gaps. either the data must contain explicit `null`s where we will disconnect, or the user tells us the disconnect threshold, or the backend response indicates expected data spacing via `field.config.interval` (which will depend on what the datasource is and what the plugin can guess from the query structure, etc).

**Comment by Brascal** [comment]
Couldn't we just make the disconnect values threshold _proportional_ to the current interval? It would fit my needs to be able to simply use something like `$__interval * 10`. That way the user would be able to get a sensible view of data gaps (proportional to their zoom level), and the graph would never disappear even if they zoomed out to 1 year or more.

**Comment by leeoniya** [comment]
we can probably interpolate `$__interval`, but adding support for expressions as suggested above (and in OP) is more involved. i'm unaware of an existing expression evaluator that we already use in the panels, which means probably integrating something like https://github.com/dy/subscript or https://github.com/justinfagnani/jexpr

**Comment by trueb2** [comment]
Even just interpolating `$__interval` is a significantly more expressive choice than a static interval. Data sources have numerous backend implementations, so imo Prometheus or FluxV2 backends aren't the core of this issue.

No matter the backend, requiring that the database/source to generate null gaps requires indicating to the data source what the `$__interval` is for the query, which is only inferred via heuristic, implemented as a specialized function, or embedded as a complicated query for some backends.

**Comment by aangelisc** [comment]
Hey @leeoniya, do you know if this is still an issue that we should attempt to solve in Influx vs one that should potentially be solved more generally?

If this is specific to Influx then I can reopen @itsmylife's PR and review it for merge. If not then I'll remove this from our board.

Thanks!

**Comment by SabrCat** [comment]
@aangelisc I hope I have ended up in the right issue here, but I'm seeing something similar. I'm using questdb with grafana and I know for absolutely certain that I am writing data at more than 1hz at all times. Yet even if I set the disconnect value threshold to 2 or 20 seconds, I still get gaps in the data that shift around depending on refresh. So in light of this being with questDB instead of influx I'd say this is a more general issue. 

https://github.com/user-attachments/assets/0a35f33d-1460-4d14-b50d-46d148c1f760

Setting the threshold to something ridiculous like 1 day fixes the issue, but I need it to be immediately visible if data has been missed even for just two or three seconds. And if I want that I get constant gaps in the graph at the moment.

**Comment by SabrCat** [comment]
With further testing I've now found out that at least in my case the gaps are caused by using ``Order by Timestamp DESC``

<img width="1023" height="257" alt="Image" src="https://github.com/user-attachments/assets/5ca12ef4-d035-4975-a0f7-de068dd6b63d" />

removing this option does fix the holes in the graph, however it no longer allows me to use the limit as I then only get the first X data points, not the most recent data points I do get with the order by set. The limit must be set as otherwise requesting too many data points from the DB causes the query to take too long for the high update rate I need.

**Comment by leeoniya** [comment]
> removing this option does fix the holes in the graph, however it no longer allows me to use the limit as I then only get the first X data points, not the most recent data points I do get with the order by set.

you can leave the DESC and LIMIT in the query, and then reverse the sort using the **Sort by** transformation:

https://grafana.com/docs/grafana/latest/panels-visualizations/query-transform-data/transform-data/#sort-by

**Comment by SabrCat** [comment]
> > removing this option does fix the holes in the graph, however it no longer allows me to use the limit as I then only get the first X data points, not the most recent data points I do get with the order by set.
> 
> you can leave the DESC and LIMIT in the query, and then reverse the sort using the **Sort by** transformation:
> 
> https://grafana.com/docs/grafana/latest/panels-visualizations/query-transform-data/transform-data/#sort-by

I just tested this, and it works! In the meantime I also found a different solution. Query as time series not as table. 
```SELECT timestamp as time, airspeed_m_s, timestamp, first(timestamp) FROM "fixedwing_metrics" WHERE $__timeFilter(timestamp) SAMPLE BY $__sampleByInterval ALIGN TO CALENDAR```
This should hopefully also only request the points actually being displayed instead of all points like the table query with no limit (which caused too slow refresh rate). What is strange though is that even if I choose time series instead of aggregate query, I still have to add a dummy aggregate first(timestamp) otherwise I get an error in the query from questDB. Also, this has the issue that if I zoom out too far, the issue mentioned in the first post of this issue of every point getting disconnected now appears.

**Comment by alyssajoyner** [comment]
Closing this as resolved.

**Comment by bobemoe** [comment]
Was this actually resolved?   

For me I have a Prometheus query (virctoria metrics) which renders just fine (gaps in all the right places... inferred from the $interval I'm guessing?) the issues happens after I apply some transforms (labels to fields, merge with a mysql query to lookup some label id to name, then prepare time series) after which the data looks exactly the same but with more labels in the column names, however all the gaps have disappeared. I think $interval has been lost somewhere along the transform chain. 

Disconnect values works to clean it up but I hit the OP issue that it only works until I zoom out.

EDIT: Ideally the interval would be preserved after I transformed so I dont even need to disconnect... am I missing something here?

**Comment by Quacken8** [comment]
Yea, this totally wasn't resolved. To me it just seems that another user came in with an unrelated problem, found an error in their query and this was interpreted by @alyssajoyner as fixing the original issue. In the current version if I select a range long enough that the threshold lies below the sampling interval the graph becomes practically a scatter plot as all values get disconnected.

Disconnect Values should either not apply when the threshold is larger that the sampling interval selected by the panel or we should be allowed to use expression and create a threshold proportional to this interval. Otherwise this feature is not usable for situations where the user can zoom out a bit

**Comment by Quacken8** [comment]
For anyone who is struggling with this the workaround I used was a combination of conditionals in my data query and leveraging the `Connect Null Values` option and leaving `Disconnect Values` set to `Never`. My internal sampling rate is every 10 seconds, so I would cap the `aggregateWindow`'s `every` at `10s` and make it create null values on empty windows. That way it would force the disconnects into existence via the null rows in any time range

```
...
  |> aggregateWindow(
     every: 
        if int(v: $__interval) < int(v: 10s) 
           then 10s 
           else $__interval, 
     fn: mean, 
     createEmpty: true
    )
...
```
